infra/src/main/kotlin/fi/fta/geoviite/infra/publication/PublicationService.kt

```diff
@@ -523,21 +522,36 @@ class PublicationService @Autowired constructor(
         val geocodingErrors = if (trackNumber.exists && referenceLine != null) {
             validateGeocodingContext(cacheKeys[version.officialId], VALIDATION_TRACK_NUMBER)
         } else listOf()
-        val duplicateNameErrors = validateTrackNumberNumberDuplication(trackNumber, version.validatedAssetVersion.id)
+        val duplicateNameErrors = validateTrackNumberNumberDuplication(trackNumber, validationVersions)
         return fieldErrors + referenceErrors + geocodingErrors + duplicateNameErrors
     }
 
     private fun validateTrackNumberNumberDuplication(
         trackNumber: TrackLayoutTrackNumber,
-        id: IntId<TrackLayoutTrackNumber>,
+        versions: ValidationVersions,
     ): List<PublishValidationError> {
-        return if (trackNumberDao.officialDuplicateNumberExistsFor(id)) listOf(
-            PublishValidationError(
-                PublishValidationErrorType.WARNING,
-                "validation.layout.track-number.duplicate-name",
+        val drafts = versions.trackNumbers.map { it.officialId to trackNumberService.get(it.validatedAssetVersion) }
+        val officials =
+            trackNumberDao.fetchVersions(OFFICIAL, false).map(trackNumberService::get).filterNot { official ->
+                drafts.map { draft -> draft.first }.contains(official.id)
+            }
+        val officialDuplicateExists =
+            officials.any { official -> official.id != trackNumber.id && official.number == trackNumber.number }
+        val stagedDuplicateExists =
+            drafts.any { (_, draft) -> draft.number == trackNumber.number && draft.id != trackNumber.id && draft.state != LayoutState.DELETED }
```

Comment:
Pihviä. Paljon hakuja, mutta lähinnä kakutettuja sellaisia -> en nähnyt toistaiseksi järkeväksi ruveta ropaamaan kutsuketjujen signatureja sen osalta, että näitä saataisiin haettua harvemmin